### ISSUE TITLE
12.2020 - High Level Goal - IBC Handlers

### ISSUE THREAD

**Issue by ebuchman** [informalsystems/ibc-rs#119]
## Summary

Implement IBC handlers and integrate in a rust-abci app. Goal is to improve testing via:

- language agnostic test vectors to test handlers in Go and Rust
- Rust IBC app that can connect to a Gaia testnet
- Rust IBC app that can be used for Rust Relayer integration testing (ie. relaying between two Rust IBC apps built on Tendermint Core)

## Proposal

- IBC data structures
  - most (all?) should be handled by #114 as they're already needed by relayer
- IBC handlers
  - [x] #159 -- client ([ICS02](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics) and [ICS07](https://github.com/cosmos/ics/tree/master/spec/ics-007-tendermint-client), though much of the ICS07 functionality should come from the tendermint-rs light client) 
  - [x] #160 -- connection ([ICS03](https://github.com/cosmos/ics/tree/master/spec/ics-003-connection-semantics))
  - [ ] #315, #316 channels and packets ([ICS04](https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics) - this is the most complex/advanced of the bunch)
  - token transfer ([ICS20](https://github.com/cosmos/ics/tree/master/spec/ics-020-fungible-token-transfer))
- Minimal ABCI Framework to support IBC (may belong in tendermint-rs once rust-abci is merged in, https://github.com/informalsystems/tendermint-rs/issues/29)
  - transaction verification (copy SDK? get construction/signing for free then)
  - storage ([ICS23](https://github.com/cosmos/ics/tree/master/spec/ics-023-vector-commitments)) - need an abstraction, do we need to implement IAVL+ or can will Gaia understand something simpler as well?
  - port binding ([ICS05](https://github.com/cosmos/ics/tree/master/spec/ics-005-port-allocation))
  - host requirements ([ICS24](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements), [ICS25](https://github.com/cosmos/ics/tree/master/spec/ics-025-handler-interface), [ICS26](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module))
  - bank module

Questions:
- How important is the Loopback handler ([ICS09](https://github.com/cosmos/ics/tree/master/spec/ics-009-loopback-client)) ?
- Can we minimize initial requirements from the ABCI framework, eg. ICS 05,23,24,25,26

**Comment by cwgoes** [comment]
Note that the SDK implementation already supports [generalised Merkle proofs](https://github.com/confio/ics23), so it shouldn't be hard to use a simpler (non-IAVL+) store if that would be easier for an initial Rust prototype.

The loopback handler is not particularly important, and bears no relevance to cross-implementation testing.

**Comment by adizere** [comment]
Moved most of the relevant issues from here into #385. For the ABCI framework, some issues are work in progress (signing), some are almost done (host requirements), while others are deferred to a later relayer version (bank module).
